src/main/java/pl/project13/maven/git/build/BuildServerDataProvider.java

```diff
@@ -0,0 +1,151 @@
+package pl.project13.maven.git.build;
+
+import org.apache.maven.project.MavenProject;
+import org.jetbrains.annotations.NotNull;
+import pl.project13.maven.git.GitCommitPropertyConstant;
+import pl.project13.maven.git.log.LoggerBridge;
+import pl.project13.maven.git.util.PropertyManager;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Map;
+import java.util.Properties;
+import java.util.TimeZone;
+
+public abstract class BuildServerDataProvider {
+
+  private final LoggerBridge log;
+  private String dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ";
+  private String dateFormatTimeZone = null;
+  private String prefixDot = "";
+  private MavenProject project = null;
+
+  BuildServerDataProvider(LoggerBridge log) {
+    this.log = log;
+  }
+
+  public BuildServerDataProvider setDateFormat(@NotNull String dateFormat) {
+    this.dateFormat = dateFormat;
+    return this;
+  }
+
+  public BuildServerDataProvider setDateFormatTimeZone(@NotNull String dateFormatTimeZone) {
+    this.dateFormatTimeZone = dateFormatTimeZone;
+    return this;
+  }
+
+  public BuildServerDataProvider setProject(@NotNull MavenProject project) {
+    this.project = project;
+    return this;
+  }
+
+  public BuildServerDataProvider setPrefixDot(@NotNull String prefixDot) {
+    this.prefixDot = prefixDot;
+    return this;
+  }
+
+  public abstract BuildEnvironmentType getBuildEnvironmentType();
+
+  /**
+   * Get the {@link BuildServerDataProvider} impelementation for the running environment
+   * @param env environment variables which get used to identify the environment
+   * @param log logging provider which will be used to log events
+   * @return the corresponding {@link BuildServerDataProvider} for your environment or {@link UnknownBuildServerData}
+   */
+  public static BuildServerDataProvider getBuildServerProvider(@NotNull Map<String, String> env, LoggerBridge log) {
+    switch (checkBuildEnvironmentForType(env)) {
```

Comment:
I believe with with one of my previous comments this would be the last reference of the ``checkBuildEnvironmentForType`` method call.

The switch and checks done in this specific function somewhat look duplicated to what is being done here.
In any case we somewhat return an instance of a BuildServerDataProvider, why not get rid of the ``checkBuildEnvironmentForType`` method and change it to:
```java
public static BuildServerDataProvider getBuildServerProvider(@NotNull Map<String, String> env, LoggerBridge log) {
    if (BambooBuildServerData.isActiveServer(env)) {
      return new BambooBuildServerData(log);
    }
    if (GitlabBuildServerData.isActiveServer(env)) {
      return new GitlabBuildServerData(log);
    }
.....
}
```

It seems that with all the rework the enum ``BuildEnvironmentType`` would also become not needed anymore (I think we still need to the distinction between an unknown environment and and actual build server, but this could be done with ``if(buildServerDataProvider instanceof UnknownBuildServerData)``

Reply:
Yes you are right we can drop the `BuildEnvironmentType` and the corresponding methods too.